[Bug]: Error: Plugin did not respond - resource: grafana_rule_group

### Terraform Version

1.8.2

### Terraform Grafana Provider Version

2.18.0

### Grafana Version

9.4

### Affected Resource(s)

* grafana_rule_group

### Terraform Configuration Files

```hcl
resource "grafana_rule_group" "this" {
  provider         = grafana.customer
  name             = local.dashboard_name
  folder_uid       = var.grafana_folder_id
  interval_seconds = 300

  rule {
    name      = "${local.dashboard_name}-DISK%"
    condition = "C"

    data {
      ref_id = "A"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "jO5EHHfIk"
      model          = "{\"accountId\":\"${data.aws_caller_identity.current.account_id}\",\"dimensions\":{\"InstanceId\":\"i-04d1cf267aa0cb8ac\"},\"expression\":\"\",\"id\":\"\",\"intervalMs\":1000,\"label\":\"\",\"logGroups\":[],\"matchExact\":true,\"maxDataPoints\":43200,\"metricEditorMode\":0,\"metricName\":\"disk_used_percent\",\"metricQueryType\":0,\"namespace\":\"CWAgent\",\"period\":\"\",\"queryMode\":\"Metrics\",\"refId\":\"A\",\"region\":\"default\",\"sqlExpression\":\"\",\"statistic\":\"Average\"}"
    }
    data {
      ref_id = "B"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "__expr__"
      model          = "{\"conditions\":[{\"evaluator\":{\"params\":[],\"type\":\"gt\"},\"operator\":{\"type\":\"and\"},\"query\":{\"params\":[\"B\"]},\"reducer\":{\"params\":[],\"type\":\"last\"},\"type\":\"query\"}],\"datasource\":{\"type\":\"__expr__\",\"uid\":\"__expr__\"},\"expression\":\"A\",\"intervalMs\":1000,\"maxDataPoints\":43200,\"reducer\":\"last\",\"refId\":\"B\",\"type\":\"reduce\"}"
    }
    data {
      ref_id = "C"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "__expr__"
      model          = "{\"conditions\":[{\"evaluator\":{\"params\":[70],\"type\":\"gt\"},\"operator\":{\"type\":\"and\"},\"query\":{\"params\":[\"C\"]},\"reducer\":{\"params\":[],\"type\":\"last\"},\"type\":\"query\"}],\"datasource\":{\"type\":\"__expr__\",\"uid\":\"__expr__\"},\"expression\":\"B\",\"intervalMs\":1000,\"maxDataPoints\":43200,\"refId\":\"C\",\"type\":\"threshold\"}"
    }

    no_data_state  = "NoData"
    exec_err_state = "Error"
    for            = "5m"
    annotations = {
      "__dashboardUid__" = grafana_dashboard.this.uid
      "__panelId__"      = "4"
    }
    is_paused = false
  }

  rule {
    name      = "${local.dashboard_name}-RAM%"
    condition = "C"

    data {
      ref_id = "A"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "jO5EHHfIk"
      model          = "{\"accountId\":\"${data.aws_caller_identity.current.account_id}\",\"dimensions\":{\"InstanceId\":\"i-04d1cf267aa0cb8ac\"},\"expression\":\"\",\"id\":\"\",\"intervalMs\":1000,\"label\":\"\",\"logGroups\":[],\"matchExact\":true,\"maxDataPoints\":43200,\"metricEditorMode\":0,\"metricName\":\"mem_used_percent\",\"metricQueryType\":0,\"namespace\":\"CWAgent\",\"period\":\"\",\"queryMode\":\"Metrics\",\"refId\":\"A\",\"region\":\"default\",\"sqlExpression\":\"\",\"statistic\":\"Average\"}"
    }
    data {
      ref_id = "B"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "__expr__"
      model          = "{\"conditions\":[{\"evaluator\":{\"params\":[],\"type\":\"gt\"},\"operator\":{\"type\":\"and\"},\"query\":{\"params\":[\"B\"]},\"reducer\":{\"params\":[],\"type\":\"last\"},\"type\":\"query\"}],\"datasource\":{\"type\":\"__expr__\",\"uid\":\"__expr__\"},\"expression\":\"A\",\"intervalMs\":1000,\"maxDataPoints\":43200,\"reducer\":\"last\",\"refId\":\"B\",\"type\":\"reduce\"}"
    }
    data {
      ref_id = "C"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "__expr__"
      model          = "{\"conditions\":[{\"evaluator\":{\"params\":[97],\"type\":\"gt\"},\"operator\":{\"type\":\"and\"},\"query\":{\"params\":[\"C\"]},\"reducer\":{\"params\":[],\"type\":\"last\"},\"type\":\"query\"}],\"datasource\":{\"type\":\"__expr__\",\"uid\":\"__expr__\"},\"expression\":\"B\",\"intervalMs\":1000,\"maxDataPoints\":43200,\"refId\":\"C\",\"type\":\"threshold\"}"
    }

    no_data_state  = "NoData"
    exec_err_state = "Error"
    for            = "5m"
    annotations = {
      "__dashboardUid__" = grafana_dashboard.this.uid
      "__panelId__"      = "2"
    }
    is_paused = false
  }

  rule {
    name      = "${local.dashboard_name}-CPU%"
    condition = "C"

    data {
      ref_id = "A"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "jO5EHHfIk"
      model          = "{\"accountId\":\"${data.aws_caller_identity.current.account_id}\",\"dimensions\":{\"InstanceId\":\"i-04d1cf267aa0cb8ac\"},\"expression\":\"\",\"id\":\"\",\"intervalMs\":1000,\"label\":\"\",\"logGroups\":[],\"matchExact\":true,\"maxDataPoints\":43200,\"metricEditorMode\":0,\"metricName\":\"CPUUtilization\",\"metricQueryType\":0,\"namespace\":\"AWS/EC2\",\"period\":\"\",\"queryMode\":\"Metrics\",\"refId\":\"A\",\"region\":\"default\",\"sqlExpression\":\"\",\"statistic\":\"Average\"}"
    }
    data {
      ref_id = "B"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "__expr__"
      model          = "{\"conditions\":[{\"evaluator\":{\"params\":[],\"type\":\"gt\"},\"operator\":{\"type\":\"and\"},\"query\":{\"params\":[\"B\"]},\"reducer\":{\"params\":[],\"type\":\"last\"},\"type\":\"query\"}],\"datasource\":{\"type\":\"__expr__\",\"uid\":\"__expr__\"},\"expression\":\"A\",\"intervalMs\":1000,\"maxDataPoints\":43200,\"reducer\":\"last\",\"refId\":\"B\",\"type\":\"reduce\"}"
    }
    data {
      ref_id = "C"

      relative_time_range {
        from = 3600
        to   = 0
      }

      datasource_uid = "__expr__"
      model          = "{\"conditions\":[{\"evaluator\":{\"params\":[70],\"type\":\"gt\"},\"operator\":{\"type\":\"and\"},\"query\":{\"params\":[\"C\"]},\"reducer\":{\"params\":[],\"type\":\"last\"},\"type\":\"query\"}],\"datasource\":{\"type\":\"__expr__\",\"uid\":\"__expr__\"},\"expression\":\"B\",\"intervalMs\":1000,\"maxDataPoints\":43200,\"refId\":\"C\",\"type\":\"threshold\"}"
    }

    no_data_state  = "NoData"
    exec_err_state = "Error"
    for            = "5m"
    annotations = {
      "__dashboardUid__" = grafana_dashboard.this.uid
      "__panelId__"      = "2"
    }
    is_paused = false
  }
}
```

### Expected Behavior

Terraform should create 3 alert rules, inside an existing folder:

- {local.dashboard_name}-DISK%
- {local.dashboard_name}-CPU%
- {local.dashboard_name}-RAM%

### Actual Behavior

```
╷
│ Error: Plugin did not respond
│ 
│   with module.ec2_resource_1.grafana_rule_group.this,
│   on .terraform/modules/ec2_resource_1/grafana_alerts.tf line 1, in resource "grafana_rule_group" "this":
│    1: resource "grafana_rule_group" "this" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-grafana_v2.18.0 plugin:

panic: interface conversion: interface {} is int64, not string

goroutine 52 [running]:
github.com/grafana/terraform-provider-grafana/v2/internal/resources/grafana.readAlertRuleGroup({0x14000192080?, 0x140005446c0?}, 0x14000192080, {0x105946b40, 0x14000068840})
        github.com/grafana/terraform-provider-grafana/v2/internal/resources/grafana/resource_alerting_rule_group.go:263 +0x628
github.com/grafana/terraform-provider-grafana/v2/internal/resources/grafana.putAlertRuleGroup({0x105bd7150, 0x1400027f810}, 0x14000068840?, {0x105946b40, 0x14000068840})
        github.com/grafana/terraform-provider-grafana/v2/internal/resources/grafana/resource_alerting_rule_group.go:335 +0x54c
github.com/grafana/terraform-provider-grafana/v2/internal/resources/grafana.addValidationToSchema.func4({0x105bd7150, 0x1400027f810}, 0x0?, {0x105946b40?, 0x14000068840})
        github.com/grafana/terraform-provider-grafana/v2/internal/resources/grafana/resources.go:68 +0x19c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x140005209a0, {0x105bd70a8, 0x14000619e90}, 0xd?, {0x105946b40, 0x14000068840})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/schema/resource.go:778 +0xe8
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x140005209a0, {0x105bd70a8, 0x14000619e90}, 0x1400061c5b0, 0x14000192180, {0x105946b40, 0x14000068840})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/schema/resource.go:909 +0x86c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x14000464ff0, {0x105bd70a8?, 0x14000619da0?}, 0x14000652f50)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.33.0/helper/schema/grpc_provider.go:1078 +0xb08
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0x105bd70e0?, {0x105bd70a8?, 0x14000619aa0?}, 0x14000652f50)
        github.com/hashicorp/terraform-plugin-mux@v0.15.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x184
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x140004268c0, {0x105bd70a8?, 0x140006190b0?}, 0x1400022f7a0)
        github.com/hashicorp/terraform-plugin-go@v0.22.1/tfprotov5/tf5server/server.go:846 +0x2b0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x105b71860?, 0x140004268c0}, {0x105bd70a8, 0x140006190b0}, 0x1400037ae00, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.22.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:518 +0x164
google.golang.org/grpc.(*Server).processUnaryRPC(0x140001c4e00, {0x105bd70a8, 0x14000619020}, {0x105bdd480, 0x14000493860}, 0x14000661b00, 0x140001c9ce0, 0x106803598, 0x0)
        google.golang.org/grpc@v1.62.1/server.go:1386 +0xba0
google.golang.org/grpc.(*Server).handleStream(0x140001c4e00, {0x105bdd480, 0x14000493860}, 0x14000661b00)
        google.golang.org/grpc@v1.62.1/server.go:1797 +0xc70
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.62.1/server.go:1027 +0x8c
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 43
        google.golang.org/grpc@v1.62.1/server.go:1038 +0x150

Error: The terraform-provider-grafana_v2.18.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Steps to Reproduce

1. Terraform apply

### Important Factoids

- Runs inside a custom Terraform module used by our company, works fine outside of this module.
- We are using AWS Managed Grafana

### References

_No response_

--- a/internal/resources/grafana/resource_alerting_rule_group.go
+++ b/internal/resources/grafana/resource_alerting_rule_group.go
@@ -55,10 +55,11 @@ This resource requires Grafana 9.1.0 or later.
 				Description: "The name of the rule group.",
 			},
 			"folder_uid": {
-				Type:        schema.TypeString,
-				Required:    true,
-				ForceNew:    true,
-				Description: "The UID of the folder that the group belongs to.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "The UID of the folder that the group belongs to.",
+				ValidateFunc: folderUIDValidation,
 			},
 			"interval_seconds": {
 				Type:        schema.TypeInt,

--- a/internal/resources/grafana/resource_folder.go
+++ b/internal/resources/grafana/resource_folder.go
@@ -3,6 +3,7 @@ package grafana
 import (
 	"context"
 	"encoding/json"
+	"regexp"
 	"strconv"
 
 	goapi "github.com/grafana/grafana-openapi-client-go/client"
@@ -13,7 +14,10 @@ import (
 	"github.com/grafana/terraform-provider-grafana/v2/internal/common"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 )
+
+var folderUIDValidation = validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9\-\_]+$`), "folder UIDs can only be alphanumeric, dashes, or underscores")
 
 func resourceFolder() *common.Resource {
 	schema := &schema.Resource{
@@ -34,11 +38,12 @@ func resourceFolder() *common.Resource {
 		Schema: map[string]*schema.Schema{
 			"org_id": orgIDAttribute(),
 			"uid": {
-				Type:        schema.TypeString,
-				Computed:    true,
-				Optional:    true,
-				ForceNew:    true,
-				Description: "Unique identifier.",
+				Type:         schema.TypeString,
+				Computed:     true,
+				Optional:     true,
+				ForceNew:     true,
+				Description:  "Unique identifier.",
+				ValidateFunc: folderUIDValidation,
 			},
 			"title": {
 				Type:        schema.TypeString,

--- a/internal/resources/grafana/resource_folder_permission.go
+++ b/internal/resources/grafana/resource_folder_permission.go
@@ -30,10 +30,11 @@ Manages the entire set of permissions for a folder. Permissions that aren't spec
 
 		Schema: map[string]*schema.Schema{
 			"folder_uid": {
-				Type:        schema.TypeString,
-				Required:    true,
-				ForceNew:    true,
-				Description: "The UID of the folder.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "The UID of the folder.",
+				ValidateFunc: folderUIDValidation,
 			},
 		},
 	}

--- a/internal/resources/grafana/resource_library_panel.go
+++ b/internal/resources/grafana/resource_library_panel.go
@@ -73,6 +73,7 @@ Manages Grafana library panels.
 					_, new = SplitOrgResourceID(new)
 					return old == new
 				},
+				ValidateFunc: folderUIDValidation,
 			},
 			"name": {
 				Type:        schema.TypeString,

